### PR TITLE
Hw23 lesson15_CartTests

### PR DESCRIPTION
**Pull request by Svitlytsya**

### CODE REVIEW

File: LambdatestEcom/Pages/ProductPage.cs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    internal class ProductPage
+    {
+        private readonly IPage _page;
+
+        public ProductPage(IPage page)
+        {
+            _page = page;
+        }
+
+        public async Task IncreaseQuantityOfProduct(int quantity)
```
Comment (vldemyan) [inline LambdatestEcom/Pages/ProductPage.cs:21]: quantity parameter is never used. Regardless what number is provided, it is always +2

File: LambdatestEcom/Pages/ProductPage.cs [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+            await _page.GetByRole(AriaRole.Button, new() { Name = "Increase quantity" }).ClickAsync();
+            await _page.GetByRole(AriaRole.Button, new() { Name = "Increase quantity" }).ClickAsync();
+            //await _page.GetByRole(AriaRole.Button).Filter(new() { HasText = "Increase quantity" }).ClickAsync();
+            //await _page.GetByRole(AriaRole.Button).Filter(new() { HasText = "Increase quantity" }).ClickAsync();
+            
+        }
+         public async Task AddProductToCart()
+         {
+            //await _page.GetByRole(AriaRole.Button, new() { Name = "Add to Cart" }).ClickAsync();
+            await _page.Locator("#entry_216842").GetByRole(AriaRole.Button, new() { Name = "Add to Cart", Exact = true }).ClickAsync();
```
Comment (vldemyan) [inline LambdatestEcom/Pages/ProductPage.cs]: Locator("#entry_216842") will work only with one product. AddProductToCart will fail for any other product

File: LambdatestEcom/Pages/ShopCartPage.cs [hunk 3]
```diff
@@ -3,10 +3,49 @@
...
+            await _page.WaitForLoadStateAsync(LoadState.DOMContentLoaded);
+        }
+
+
+        public async Task EditProductQuantityinCart(int quantity)
+        {
+            //await _page.Locator("#content").GetByRole(AriaRole.Table).GetByRole(AriaRole.Row).Filter(new() { HasText = "iPod Classic" }).GetByRole(AriaRole.Cell, new() { Name = "1" }).ClickAsync();
+            //await _page.Locator("#content").GetByRole(AriaRole.Table).GetByRole(AriaRole.Row).Filter(new() { HasText = "iPod Classic" }).GetByRole(AriaRole.Cell, new() { Name = "1" }).FillAsync(quantity.ToString());
+            //await _page.Locator("input[name=\"quantity\\[88149\\]\"]").FillAsync("2");
+            await _page.Locator("#content").GetByRole(AriaRole.Table).GetByRole(AriaRole.Row).Filter(new() { HasText = "iPod Classic" }).Locator("input[name^='quantity']").FillAsync(quantity.ToString());
```
Comment (vldemyan) [inline LambdatestEcom/Pages/ShopCartPage.cs]: "iPod Classic" should be parameter for this function but not hardcoded here

File: LambdatestEcom/Pages/ShopCartPage.cs [hunk 4]
```diff
@@ -3,10 +3,49 @@
...
+        {
+            await _page.GetByRole(AriaRole.Link, new() { Name = "Continue Shopping" }).ClickAsync();
+            await _page.WaitForLoadStateAsync(LoadState.NetworkIdle);
+            await _page.WaitForLoadStateAsync(LoadState.DOMContentLoaded);
+        }
+
+
+
+
+        public async Task UpdateQuantity()
```
Comment (vldemyan) [inline LambdatestEcom/Pages/ShopCartPage.cs]: Name UpdateQuantity looks incorrect. What it does? Increase quantity?

File: LambdatestEcom/Tests/CartTests.cs [hunk 5]
```diff
@@ -1,12 +1,69 @@
...
+        {
+            // Arrange
+            DateTime now = DateTime.Now;
+            string randomString = now.ToString("yyyyMMddHHmmss");
+
+            var homePage = new HomePage(page);
+            var catalogPage = new CatalogPage(page);
+            var productPage = new ProductPage(page);
+            var shopCartPage = new ShopCartPage(page);
+            var addedProductRow = page.Locator("#content")
```
Comment (vldemyan) [inline LambdatestEcom/Tests/CartTests.cs]: locators must not be in the test. You need one function in ShopCartPage, like GetAddedProductLocator(string productName)
so you can use it like 
var addedProductRow = GetAddedProductLocator("HP LP3065");
var addedProductRow2 = GetAddedProductLocator("iPod Classic");

File: LambdatestEcom/Tests/CartTests.cs [hunk 6]
```diff
@@ -1,12 +1,69 @@
...
+            await homePage. SearchProductBySearchButton("iPod Classic");
+            await catalogPage.AddProductToCart(0);
+            await productPage.GoToShoppingCart();
+
+            await Assertions.Expect(addedProductRow).ToContainTextAsync("3");
+            await Assertions.Expect(addedProductRow2).ToContainTextAsync("1");
+            await Assertions.Expect(page.Locator("#content")).ToContainTextAsync("$488.00");
+
+            await shopCartPage.EditProductQuantityinCart(2);
+            await shopCartPage.UpdateQuantity();
```
Comment (vldemyan) [inline LambdatestEcom/Tests/CartTests.cs]: if this UpdateQuantity is needed after EditProductQuantityinCart then just put code of UpdateQuantity at the end of EditProductQuantityinCart  and remove UpdateQuantity() at all as it does not make any sense to have it as separate function

File: LambdatestEcom/Pages/ProductPage.cs (hunk 1 above)
Comment (Svitlytsya) [reply LambdatestEcom/Pages/ProductPage.cs:21]: Thanks, fixed.

File: LambdatestEcom/Pages/ProductPage.cs (hunk 2 above)
Comment (Svitlytsya) [reply LambdatestEcom/Pages/ProductPage.cs]: Ok, thanks, fixed.

File: LambdatestEcom/Pages/ShopCartPage.cs (hunk 3 above)
Comment (Svitlytsya) [reply LambdatestEcom/Pages/ShopCartPage.cs]: Thanks, fixed

File: LambdatestEcom/Pages/ShopCartPage.cs (hunk 4 above)
Comment (Svitlytsya) [reply LambdatestEcom/Pages/ShopCartPage.cs]: Ок;) understand..!

File: LambdatestEcom/Tests/CartTests.cs (hunk 6 above)
Comment (Svitlytsya) [reply LambdatestEcom/Tests/CartTests.cs]: Ok, thanks, done.

File: LambdatestEcom/Tests/CartTests.cs (hunk 5 above)
Comment (Svitlytsya) [reply LambdatestEcom/Tests/CartTests.cs]: Thanks, done.

File: LambdatestEcom/Pages/ProductPage.cs [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+    internal class ProductPage
+    {
+        private readonly IPage _page;
+
+        public ProductPage(IPage page)
+        {
+            _page = page;
+        }
+
+        public async Task IncreaseQuantityOfProduct()
```
Comment (vldemyan) [inline LambdatestEcom/Pages/ProductPage.cs]: by "parameter is not used" I meant that you need to use it. IncreaseQuantityOfProduct can't be hardcoded to always increase by 2, it should be int parameter and according to this parameter, increase should be done, so you can use IncreaseQuantityOfProduct(1), IncreaseQuantityOfProduct(2), IncreaseQuantityOfProduct(3)

File: LambdatestEcom/Pages/ShopCartPage.cs [hunk 8]
```diff
@@ -3,10 +3,54 @@
...
+        //    await _page.GetByRole(AriaRole.Button, new() { Name = "" }).Nth(1).ClickAsync();
+        //    await _page.WaitForLoadStateAsync(LoadState.NetworkIdle);
+        //    await _page.WaitForLoadStateAsync(LoadState.DOMContentLoaded);
+        //}
+
+
+        public async Task EditProductQuantityinCart(string name, int quantity)
+        {
+            //await _page.Locator("#content").GetByRole(AriaRole.Table).GetByRole(AriaRole.Row).Filter(new() { HasText = name }).Locator("input[name^='quantity']").FillAsync(quantity.ToString());
+            await _page.Locator("#content").GetByRole(AriaRole.Table).GetByRole(AriaRole.Row).Filter(new() { HasText = name }).Locator("input[name='quantity']").FillAsync(quantity.ToString());
```
Comment (vldemyan) [inline LambdatestEcom/Pages/ShopCartPage.cs]: now you have GetAddedProductLokator, so you can replace 
_page.Locator("#content").GetByRole(AriaRole.Table).GetByRole(AriaRole.Row).Filter(new() { HasText = name })
to 
GetAddedProductLokator(name)

File: LambdatestEcom/Pages/ProductPage.cs (hunk 7 above)
Comment (Svitlytsya) [reply LambdatestEcom/Pages/ProductPage.cs]: Maybe I finally understood... but I'm not sure..;)

File: LambdatestEcom/Pages/ShopCartPage.cs (hunk 8 above)
Comment (Svitlytsya) [reply LambdatestEcom/Pages/ShopCartPage.cs]: Here, not fully understood... question in the name of the string parameter..? that is, "productName" should be replaced by "name.."? if so, done..;)

Comment (vldemyan) [reply LambdatestEcom/Pages/ShopCartPage.cs]: await _page.Locator("#content").GetByRole(AriaRole.Table).GetByRole(AriaRole.Row).Filter(new() { HasText = name }).Locator("input[name='quantity']").FillAsync(quantity.ToString());
can be replaced with
await GetAddedProductLokator(name).Locator("input[name='quantity']").FillAsync(quantity.ToString());

Comment (Svitlytsya) [reply LambdatestEcom/Pages/ShopCartPage.cs]: ок, thank you so much..!

**Review by vldemyan — APPROVED** [review]

**Pull request merged**